### ISSUE TITLE
Option to build models with Jax/Flax/Numpyro

### ISSUE THREAD

**Issue by adamgayoso** [scverse/scvi-tools#1366]
Jax-based scVI is ~5x faster than the current implementation on a GPU (notebook [here](https://colab.research.google.com/drive/1fBYpRxIO-YXOqpU2npA8p5sKYY6UaJEI?usp=sharing), versus [this](https://docs.scvi-tools.org/en/latest/tutorials/notebooks/harmonization.html) tutorial). 

We should build some backend components so that one can develop a Jax-based model and still derive from our `BaseModelClass` and use our dataloaders.

From what I understand, we need to build:

1.  `FlaxBaseModuleClass` that wraps a Flax `nn.Module` (`model.module.flax_module`) and keeps track of the optimized parameters/state (e.g., `model.module.params` , `model.module.state`).
2. We need some sort of abstract Flax nn module so that everyone organizes this the same way (e.g., inference/generative on the pytorch side)
3. A `FlaxTrainingPlan` that roughly follows the pytorch lightning analog
4. A black box trainer
5. Numpyro extensions.

To get this started we should build `scvi.model.JaxSCVI`. We should start with doing (1) more robustly, and then continuously upgrading (3) and (4). 

Some aspects, like differential expression should work easily with `JaxSCVI` because it only cares about numpy arrays. Things like scArches will have to be completely redone. Though I think in the scArches case it might be easier as we can create separate `nn.Dense` layers for the categories and add them together at no real speed cost (this was slow in pytorch).

For (5) we can use https://github.com/YosefLab/scvi-tools/pull/1146 as a starting point

And finally for (1) generally we can use `numpyro.dist` just as we do PyTorch dists.

### `JaxModuleWrapper`
- interacts with top-level scvi-tools models, has state_dict so save and load work
- has :attr:`flax_module` which is an instance of `JaxBaseModuleClass`

### `JaxBaseModuleClass`
- has inference/generative/call
- organizes loss generally like our current PyTorch `BaseModuleClass`

### `JaxTrainingPlan`
- train_step/test_step/... only interacts with `JaxBaseModuleClass`

### `JaxTrainer`
- lightweight mirror of pytorch lightning trainer

**Comment by adamgayoso** [comment]
Let's also explore if we can use Elegy in the same way we use pytorch lightning https://github.com/poets-ai/elegy

**Comment by adamgayoso** [comment]
Example repo for the level of complexity our training loop will need:

https://github.com/ncullen93/torchsample

**Comment by jjhong922** [comment]
Closing this issue since merging https://github.com/scverse/scvi-tools/pull/1506 wraps up the first drafts of the major components requested in this issue. I will split NumPyro support off into its own issue.
